src/pkg/storage/storage.go

```diff
@@ -92,3 +93,79 @@ func orEmptyString(v any) string {
 
 	return v.(string)
 }
+
+func (s Storage) StorageConfig() (Configurer, error) {
```

Comment:
I'm dubious about the value of this function.  Noting how you're required to cast back to the service config struct in all locations of its usage, I am left with the expectation that:
1. All callers will need to have _a priori_ knowledge of which provider is in use before calling this func anyway.
2. Returning a `Configurer` interface isn't helping the callers, anyway, because the data within the config is their end goal, not the behavior of the configurer.

I think a different approach is in order.  I think sticking with the original `func (s Storage) S3Config() (S3Config, error)` will have more mileage here, purely due to 2/.  It will require callers to know which provider to request at the time of usage.  But that's already the pattern in place, a pattern that isn't changing as part of this PR.

Othewise you could go really big and ditch the provider-specific structs altogether in favor of a map (or map wrapper).  That'll let you more directly focus on what _data_ is being passed around, instead of the behavior built in to the data structure.

Reply:
I think we should look for an acceptable middle-ground solution for the moment. The current goal is not to refactor all the config stuff, though there's already been a significant amount of that to enable local storage. Do we need to set time aside to do the refactor? Absolutely, but for the moment let's limit the scope of changes so we can finish up this feature

Looking at this PR and the next one, the reason we have this is so that callers can get a handle to the specific config for the provider without us having to create `ConfigForX` for every provider we have. Eventually we'll need to determine what solution we want to implement, but for now I think this is an alright tradeoff to make

If we want to make larger changes to how all this is handled I think we should come up with a plan that we agree on. This code can be either easy to handle in the future or very painful depending on what solution we pick so we should think carefully before changing things

Reply:
> Looking at this PR and the next one, the reason we have this is so that callers can get a handle to the specific config for the provider without us having to create ConfigForX for every provider we have.

The difference between the two is reduced to:
```
// ConfigForX
func buildXConfig(base map[string]string) (XConfig, error) {...}

func XConfig() (Xconfig, error) {
  return buildXConfig(m)
} 

myCfg, err := XConfig()

// proposed
func buildXConfig(base map[string]string) (XConfig, error) {...}

func StorageConfig() (configurer, error) {
  return buildXConfig(m)
} 

iCfg, err := StorageConfig()

myConfig, ok := iCfg.(*Xconfig)
```

My point in writing that out is to show that the difference here is moot.  Interfaces cannot solve all problems and, unfortunately, this is one of those cases.  What we need here is not described by shared behavior.  And, sincerely, I'm not trying to make a soapbox out of it.  If I could come up with an interface which got around it (within a reasonable amount of work), I would go for that path.  But especially as we're looking for a more direct approach to the problem with less refactoring, I think the best course here is to accept that an interface is more abstraction than is beneficial, and to keep with the straightforward data production.  However inelegant that solution may seem.

Reply:
> Othewise you could go really big and ditch the provider-specific structs altogether in favor of a map (or map wrapper)

That was my first instinct while doing this mini config refactor. However, I found that it may end up disturbing a lot of code. I am going to pass on this for now. 

> Returning a Configurer interface isn't helping the callers, anyway, because the data within the config is their end goal, not the behavior of the configurer.

This is partially correct. It's true that callers are interested in data within the config. However, in several cases the callers pass this `Configurer` interface to `WriteRepoConfig`. See [example](https://github.com/alcionai/corso/blob/03ed65a0c00c159a91897b3dc8aa365199bbaaec/src/cli/repo/s3.go#L152). You could argue here that we could ditch this in favor of a config map that is fed to viper (and eventually written to config). 
But `WriteRepoConfig` is called from several places and it's a tangled mess. Unraveling this will push back the local storage work out even further. 

So can we keep the code in this PR as is, unless you have strong reservations against it @ryanfkeepers ? :)

Reply:
I didn't do a great job explaining last friday, so let me retry. Consider this [example](https://github.com/alcionai/corso/blob/e50fe44d8a96b86e27f03998455c87e35aa76651/src/cli/utils/utils.go#L108) from the next PR in this chain. 

Here, I am no longer casting interface to a struct i.e. `AccountConnectAndWriteRepoConfig` and `WriteRepoConfig` are both agnostic to the storage provider. Here I need `StorageConfig()`, with the current setup, unless i 

> I'd like to follow up on the design discussion outside of this PR. But it's fine for now.

Just to be on the same page, since we've had a ton of discussion here, are you suggesting we do below for every storage provider,
```
func XConfig() (Xconfig, error) {
  return buildXConfig(m)
} 

```
instead of 
```
iCfg, err := StorageConfig()
myConfig, ok := iCfg.(*Xconfig)
``` ? 

Is this the only suggestion here, or are there more?

Reply:
> I didn't do a great job explaining last friday, so let me retry. Consider this [example](https://github.com/alcionai/corso/blob/e50fe44d8a96b86e27f03998455c87e35aa76651/src/cli/utils/utils.go#L108) from the next PR in this chain.

You did a fine job explaining.  The linked code is great works exactly like we'd want, given that it's acting within in a provider-agnostic space.

> Just to be on the same page, since we've had a ton of discussion here, are you suggesting we do below for every storage provider,

Yes, that is exaclty my request.  At least, for cases where we would otherwise cast from the interface to the struct.  We seem to have a generic use case for `StorageConfig`, which is a great case for it, and we should keep that around.  But in non-generic cases I don't believe we should be using it.